### PR TITLE
[DONT REVIEW] Enable in-place WGRAD accumulation for MXFP8Linear

### PR DESCRIPTION
**Pull request by anijain2305**
Human note polished by agent.

**Note - it cannot land until pytorch/pytorch #194434 is merged**

Problem
-------

In PP, each microbatch backward produces a weight gradient, and those
gradients have to be accumulated before FSDP's final reduce-scatter.

Today that accumulation happens in one of two places:

- Autograd's AccumulateGrad node, which adds each new gradient into
  ``param.grad``.
- FSDP, which takes over when the mixed precision policy's
  ``reduce_dtype`` does not match ``param.grad.dtype``.

Both spend an extra ``aten.add`` per microbatch. This PR folds that add
into the ``scaled_mm`` that already computes the WGRAD. Only the first
case is supported; the second is rejected, see Limitation below.

Implementation
--------------

MXFP8Linear already owns an autograd.Function, so the WGRAD can call
``F.scaled_addmm_`` directly (pytorch/pytorch#195409).

With ``inplace_wgrad_accum`` on:

- The first backward of an unshard window produces the gradient with
  ``F.scaled_mm``, as before.
- Every later one folds into that same buffer with ``F.scaled_addmm_``,
  so the accumulation happens in the GEMM epilogue and the separate add
  kernel disappears.
- It then clears ``parameter.grad`` and returns the buffer, so
  AccumulateGrad reattaches it instead of adding it to itself.

Limitation
----------

This does not work with torchtitan's default
``mixed_precision_reduce = float32``. When the reduce dtype differs from
the gradient dtype, FSDP moves the gradient off the parameter into its
own reduce-dtype accumulator after every microbatch, leaving nothing to
fold into, so the option would silently do nothing. The unshard hook
rejects that combination instead.

Matching the two is not this layer's call: reduce-scatter requires a
uniform gradient dtype across the group, so every parameter under the
same ``fully_shard`` call would have to move together. So this is useful
today only for FSDP + PP with a BF16 reduce dtype.

### CODE REVIEW

File: torchtitan/components/quantization/mxfp8/linear.py [hunk 1]
```diff
@@ -399,6 +458,23 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
             # the weight changes each optimizer step; inference does not.
             # TODO(anijain2305): key the operands on the parameter's
             # version counter so a frozen weight is quantized once.
+        weight_ref = None
+        if self.inplace_wgrad_accum:
+            # Dynamo sets is_compiling; GraphTrainer's make_fx tracer does not,
+            # so ask the proxy mode as well or tracing slips through.
```
Comment (tianyu-l) [inline torchtitan/components/quantization/mxfp8/linear.py:464]: So none of Dynamo or GraphTrainer supports this feature?

File: torchtitan/components/quantization/mxfp8/linear.py [hunk 2]
```diff
@@ -307,6 +344,27 @@ class MXFP8Linear(Linear):
     class Config(Linear.Config):
         """Drop-in replacement for ``Linear.Config``."""
 
+        inplace_wgrad_accum: bool = False
```
Comment (tianyu-l) [inline torchtitan/components/quantization/mxfp8/linear.py:347]: this flag may not be associated with PP
- Today's spmd gradient accumulation indeed does all-gather and reduce-scatter for every microbatch. But from @fegin 's study, if we instead do PP-style of "all-gather in first microbatch, reduce-scatter in last microbatch", this optimization would also apply.
- PP has a flag controlling how many microbatches you'd avoid RAF, so even under PP it's not always useful.
- In eager, what if we always turn it on, even when RAF=True + multiple microbatches. Would it cause much additional overhead? It sounds minimal to me.
- Then the final question is on tracing -- what's our plan over there. Would we require graph passes to enable fusion? That should be the only reason that we make this configurable.

File: torchtitan/components/quantization/mxfp8/linear.py [hunk 3]
```diff
@@ -399,6 +458,23 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
...
+                    "it reads and clears parameter.grad during backward, an "
+                    "autograd ownership protocol that is invisible to a graph. "
+                    "Under tracing the backward runs once, when parameter.grad "
+                    "is still None, so the first-contribution branch is frozen "
+                    "into the graph and the accumulation never happens."
+                )
+            # Weak so the graph does not keep the parameter alive; backward
+            # rechecks it, since a released weight means nothing to accumulate
+            # into.
+            weight_ref = weakref.ref(weight_NK)
```
Comment (tianyu-l) [inline torchtitan/components/quantization/mxfp8/linear.py:477]: could you add more notes on this `weakref`, I couldn't parse the short-info here

File: torchtitan/components/quantization/mxfp8/linear.py (hunk 2 above)
Comment (fegin) [reply torchtitan/components/quantization/mxfp8/linear.py]: I'm working on the PR to enable "all-gather in first microbatch, reduce-scatter in last microbatch" for non-PP. We can make this work with non-PP case after that.

Comment (anijain2305) [reply torchtitan/components/quantization/mxfp8/linear.py]: @fegin Can you forward me the PR? I will resume this work after your work lands.

Comment (anijain2305) [reply torchtitan/components/quantization/mxfp8/linear.py]: > Then the final question is on tracing -- what's our plan over there. Would we require graph passes to enable fusion? That should be the only reason that we make this configurable.

There are two things here - per-layer torch.compile and GraphTrainer.

*tldr - We don't need the `inplace_wgrad_accum` config flag to support compile or graph trainer. A graph pass is a better fit there.*

**torch.compile**  - Very hard to support. In the backward, we have a unsafe side effect `weight.grad = None`. In compile, we can replay side effects only for the forward, there is no way to replay side effects in the backward. So, this does not work. Nobody in compiler team is or plans to work on this.

To not error out, we can simply have `torch.is_compiling` and skip the wgrad fusion for compile.  If inductor wants, it can do a graph pass to do the fusion.

**Graph Trainer** - GraphTrainer PP integration uses `torch.autograd.grad` to construct the backward of the stages. So, there is no accumulation going on. Tracing will always choose the unfused path, and therefore GraphTrainer can also write a graph pass to fuse this.
